Content.Shared/Doors/Systems/SharedDoorSystem.Airlock.cs

```diff
@@ -0,0 +1,189 @@
+        if (_timing.ApplyingState)
+            return;
+
+        // Only show the maintenance panel if the airlock is closed
+        if (TryComp<WiresPanelComponent>(airlock, out var wiresPanel))
+        {
+            _wiresSystem.ChangePanelVisibility(airlock,
+                wiresPanel,
+                airlock.Comp.OpenPanelVisible || args.State != DoorState.Open);
+        }
+
+        // If the door is closed, we should look if the bolt was locked while closing
+        UpdateAutoClose(airlock);
+
+        // Make sure the airlock auto closes again next time it is opened
+        if (args.State != DoorState.Closed)
+            return;
+
+        airlock.Comp.AutoClose = true;
+        Dirty(airlock);
+    }
+
+    private void OnBoltsChanged(Entity<AirlockComponent> airlock, ref DoorBoltsChangedEvent args)
+    {
+        // If unbolted, reset the auto close timer
+        if (args.BoltsDown)
+            return;
+
+        UpdateAutoClose(airlock);
+    }
+
+    private void OnBeforeDoorOpened(Entity<AirlockComponent> airlock, ref BeforeDoorOpenedEvent args)
+    {
+        if (CanChangeState(airlock))
+            return;
+
+        args.Cancel();
+    }
+
+    private void OnBeforeDoorDenied(Entity<AirlockComponent> airlock, ref BeforeDoorDeniedEvent args)
+    {
+        if (CanChangeState(airlock))
+            return;
+
+        args.Cancel();
+    }
+
+    private void OnGetPryMod(EntityUid uid, AirlockComponent component, ref GetPryTimeModifierEvent args)
+    {
+        if (component.Powered)
+            args.PryTimeModifier *= component.PoweredPryModifier;
+
+        if (DoorSystem.IsBolted(uid))
+            args.PryTimeModifier *= component.BoltedPryModifier;
+    }
+
+    /// <summary>
+    /// Updates the auto close timer.
+    /// </summary>
+    public void UpdateAutoClose(Entity<AirlockComponent> airlock, DoorComponent? door = null)
+    {
+        if (!Resolve(airlock, ref door))
+            return;
+
+        if (door.State != DoorState.Open)
+            return;
+
+        if (!airlock.Comp.AutoClose)
+            return;
+
+        if (!CanChangeState(airlock))
+            return;
+
+        var autoCloseEvent = new BeforeDoorAutoCloseEvent();
+        RaiseLocalEvent(airlock, autoCloseEvent);
+
+        if (autoCloseEvent.Cancelled)
+            return;
+
+        DoorSystem.SetNextStateChange((airlock, door),
+            airlock.Comp.AutoCloseDelay * airlock.Comp.AutoCloseDelayModifier);
+    }
+
+    private void OnBeforePry(EntityUid uid, AirlockComponent component, ref BeforePryEvent args)
+    {
+        if (args.Cancelled)
+            return;
+
+        if (!component.Powered || args.PryPowered)
+            return;
+
+        args.Message = "airlock-component-cannot-pry-is-powered-message";
+
+        args.Cancelled = true;
+    }
+
+    public void UpdateEmergencyLightStatus(EntityUid uid, AirlockComponent component)
```

Comment:
```suggestion
    private void UpdateEmergencyLightStatus(EntityUid uid, AirlockComponent component)
```
IMO this is an important one so that things can't arbitrarily change the EA lights without also actually setting EA.